File: Pathfinder_1_13/src/main/java/com/github/ysl3000/bukkit/pathfinding/craftbukkit/v1_13_R1/entity/CraftInsentient.java
package com.github.ysl3000.bukkit.pathfinding.craftbukkit.v1_13_R1.entity;

import com.github.ysl3000.bukkit.pathfinding.craftbukkit.v1_13_R1.pathfinding.CraftNavigation;
import com.github.ysl3000.bukkit.pathfinding.craftbukkit.v1_13_R1.pathfinding.CraftPathfinderGoalWrapper;
import com.github.ysl3000.bukkit.pathfinding.entity.Insentient;
import net.minecraft.server.v1_13_R1.ControllerLook;
import net.minecraft.server.v1_13_R1.EntityInsentient;
import net.minecraft.server.v1_13_R1.PathfinderGoal;
import net.minecraft.server.v1_13_R1.PathfinderGoalSelector;
import org.bukkit.Location;
import org.bukkit.craftbukkit.v1_13_R1.entity.*;
import org.bukkit.entity.*;
import org.bukkit.util.Vector;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by ysl3000
 */
public class CraftInsentient implements Insentient {


    private static Method reset = null;

    static {

        try {
            reset = EntityInsentient.class.getDeclaredMethod("n");
            reset.setAccessible(true);
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        }
    }


    private Map<com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal, PathfinderGoal> nmsGoals = new HashMap<>();
    private Map<com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal, PathfinderGoal> nmsTargetGoals = new HashMap<>();


    private EntityInsentient handle;

    private com.github.ysl3000.bukkit.pathfinding.pathfinding.Navigation navigation;

    public CraftInsentient(Flying flying) {
        this(((CraftFlying) flying).getHandle());
    }

    private CraftInsentient(EntityInsentient handle) {
        this.handle = handle;
        this.navigation = new CraftNavigation(handle.getNavigation());

    }


    public CraftInsentient(EnderDragon enderDragon) {
        this(((CraftEnderDragon) enderDragon).getHandle());
    }

    public CraftInsentient(Creature creature) {
        this(((CraftCreature) creature).getHandle());
    }

    public CraftInsentient(Ambient ambient) {
        this(((CraftAmbient) ambient).getHandle());
    }

    public CraftInsentient(Slime slime) {
        this(((CraftSlime) slime).getHandle());
    }

    @Override
    public void addPathfinderGoal(int priority, com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal pathfinderGoal) {
        CraftPathfinderGoalWrapper goalWrapper = new CraftPathfinderGoalWrapper(pathfinderGoal);
        this.nmsGoals.put(pathfinderGoal, goalWrapper);
        handle.goalSelector.a(priority, goalWrapper);
    }

    @Override
    public void removePathfinderGoal(com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal pathfinderGoal) {
        if (nmsGoals.containsKey(pathfinderGoal)) {
            PathfinderGoal nmsGoal = nmsGoals.remove(pathfinderGoal);
            handle.goalSelector.a(nmsGoal);
        }
    }

    @Override
    public boolean hasPathfinderGoal(com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal pathfinderGoal) {
        return nmsGoals.containsKey(pathfinderGoal);
    }

    @Override
    public void clearPathfinderGoals() {
        handle.goalSelector = new PathfinderGoalSelector(handle.getWorld().methodProfiler);
        nmsGoals.clear();
    }


    @Override
    public void addTargetPathfinderGoal(int priority, com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal pathfinderGoal) {
        CraftPathfinderGoalWrapper goalWrapper = new CraftPathfinderGoalWrapper(pathfinderGoal);
        this.nmsTargetGoals.put(pathfinderGoal, goalWrapper);
        handle.targetSelector.a(priority, goalWrapper);
    }

    @Override
    public void removeTargetPathfinderGoal(com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal pathfinderGoal) {
        if (nmsTargetGoals.containsKey(pathfinderGoal)) {
            PathfinderGoal nmsGoal = nmsTargetGoals.remove(pathfinderGoal);
            handle.goalSelector.a(nmsGoal);
        }
    }

    @Override
    public boolean hasTargetPathfinderGoal(com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal pathfinderGoal) {
        return nmsTargetGoals.containsKey(pathfinderGoal);
    }

    @Override
    public void clearTargetPathfinderGoals() {
        handle.targetSelector = new PathfinderGoalSelector(handle.getWorld().methodProfiler);
        nmsTargetGoals.clear();
    }

    @Override
    public void jump() {
        handle.getControllerJump().a();
    }

    /**
     * The entity will look to the given location
     *
     * @param location the entity should look to
     */
    @Override
    public void lookAt(Location location) {
        handle.getControllerLook().a(location.getX(), location.getY(), location.getZ(), location.getYaw(), location.getPitch());
    }

    /**
     * The entity will look to the given entity
     *
     * @param entity the entity which is targeted with eyes
     */
    @Override
    public void lookAt(Entity entity) {
        lookAt(entity.getLocation());
    }

    @Override
    public Location getLookingAt() {
        ControllerLook controllerLook = handle.getControllerLook();
        return new Location(handle.getBukkitEntity().getWorld(), controllerLook.e(), controllerLook.f(), controllerLook.g());
    }

    @Override
    public void setMovementDirection(Vector direction, double speed) {
        handle.getControllerMove().a(direction.getX(), direction.getBlockY(), direction.getZ(), speed);
    }

    @Override
    public void setStrafeDirection(float forward, float sideward) {
        handle.getControllerMove().a(forward, sideward);
    }

    @Override
    public com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoalSelector getTargetSelector() {
        return new TargetGoalSelector();
    }

    @Override
    public com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoalSelector getGoalSelector() {
        return new GoalSelector();
    }

    @Override
    public void resetGoalsToDefault() {
        if (reset == null) return;
        try {
            reset.invoke(handle);
        } catch (IllegalAccessException | InvocationTargetException e) {
            e.printStackTrace();
        }
    }

    @Override
    public com.github.ysl3000.bukkit.pathfinding.pathfinding.ControllerJump getControllerJump() {
        return new ControllerJump();
    }

    @Override
    public com.github.ysl3000.bukkit.pathfinding.pathfinding.ControllerLook getControllerLook() {
        return new ControllerLookImpl();
    }

    @Override
    public com.github.ysl3000.bukkit.pathfinding.pathfinding.ControllerMove getControllerMove() {
        return new ControllerMove();
    }

    @Override
    public com.github.ysl3000.bukkit.pathfinding.pathfinding.Navigation getNavigation() {
        return navigation;
    }

    @Override
    public float getHeadHeight() {
        return handle.getHeadHeight();
    }

    @Override
    public int getDefaultYaw() {
        return handle.L();
    }

    @Override
    public int getDefaultPitch() {
        return handle.K();
    }

    @Override
    public boolean hasPositionChanged() {
        return handle.positionChanged;
    }

    @Override
    public void onEntityKill(LivingEntity livingEntity) {
        handle.b(livingEntity == null ? null : ((CraftLivingEntity) livingEntity).getHandle());
    }

    @Override
    public Entity getBukkitEntity() {
        return handle.getBukkitEntity();
    }

    @Override
    public void moveRelative(double motionX, double motionY, double motionZ) {
        moveRelative(motionX, motionY, motionZ, 1.0);
    }

    @Override
    public void moveRelative(double motionX, double motionY, double motionZ, double speed) {
        this.handle.motX += (motionX * speed);
        this.handle.motY += (motionY * speed);
        this.handle.motZ += (motionZ * speed);
        this.handle.recalcPosition();
    }

    @Override
    public void setRotation(float yaw, float pitch) {
        this.handle.yaw = yaw;
        this.handle.pitch = pitch;
    }

    @Override
    public void updateRenderAngles() {
        handle.getControllerMove().a();
    }


    private class GoalSelector implements com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoalSelector {
        /**
         * Add a pathfinder goal to the entity
         *
         * @param priority the priority 0 highest
         * @param goal     The goal to add
         */
        @Override
        public void addPathfinderGoal(int priority, com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal goal) {
            CraftInsentient.this.addPathfinderGoal(priority, goal);
        }

        /**
         * Remove a pathfinder goal from the entity
         *
         * @param goal The goal to remove
         */
        @Override
        public void removePathfinderGoal(com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal goal) {
            CraftInsentient.this.removePathfinderGoal(goal);
        }

        /**
         * Will clear all goals
         */
        @Override
        public void clearGoals() {
            CraftInsentient.this.clearPathfinderGoals();
        }
    }

    private class TargetGoalSelector implements com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoalSelector {
        /**
         * Add a pathfinder goal to the entity
         *
         * @param priority the priority 0 highest
         * @param goal     The goal to add
         */
        @Override
        public void addPathfinderGoal(int priority, com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal goal) {
            CraftInsentient.this.addTargetPathfinderGoal(priority, goal);
        }

        /**
         * Remove a pathfinder goal from the entity
         *
         * @param goal The goal to remove
         */
        @Override
        public void removePathfinderGoal(com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal goal) {
            CraftInsentient.this.removeTargetPathfinderGoal(goal);
        }

        /**
         * Will clear all goals
         */
        @Override
        public void clearGoals() {
            CraftInsentient.this.clearTargetPathfinderGoals();
        }
    }

    private class ControllerJump implements com.github.ysl3000.bukkit.pathfinding.pathfinding.ControllerJump {

        /**
         * Lets the entity jump
         */
        @Override
        public void jump() {
            CraftInsentient.this.jump();
        }
    }

    private class ControllerMove implements com.github.ysl3000.bukkit.pathfinding.pathfinding.ControllerMove {

        @Override
        public boolean isOperationMove() {
            return handle.getControllerMove().b();
        }

        @Override
        public void move(double motionX, double motionY, double motionZ, double speed) {
            CraftInsentient.this.setMovementDirection(new Vector(motionX, motionY, motionZ), speed);
        }

        @Override
        public void move(float forward, float sideward) {
            CraftInsentient.this.setStrafeDirection(forward, sideward);
        }

        @Override
        public void update() {
            handle.getControllerMove().a();
        }

        @Override
        public double getX() {
            return handle.getControllerMove().c();
        }

        @Override
        public double getY() {
            return handle.getControllerMove().d();
        }

        @Override
        public double getZ() {
            return handle.getControllerMove().e();
        }

        @Override
        public double getSpeed() {
            return handle.getControllerMove().f();
        }
    }

    private class ControllerLookImpl implements com.github.ysl3000.bukkit.pathfinding.pathfinding.ControllerLook {

        /**
         * The entity will look to the given coordinates
         *
         * @param x     x-location
         * @param y     y-location
         * @param z     z-location
         * @param yaw   head rotation yaw
         * @param pitch head rotation pitch
         */
        @Override
        public void lookAt(double x, double y, double z, float yaw, float pitch) {
            CraftInsentient.this.lookAt(new Location(handle.getBukkitEntity().getWorld(), x, y, z, yaw, pitch));
        }

        /**
         * The entity will look to the given location
         *
         * @param location the entity should look to
         */
        @Override
        public void lookAt(Location location) {
            CraftInsentient.this.lookAt(location);
        }

        /**
         * The entity will look to the given entity
         *
         * @param entity the entity which is targeted with eyes
         * @param yaw    head rotation yaw
         * @param pitch  head rotation pitch
         */
        @Override
        public void lookAt(Entity entity, float yaw, float pitch) {
            CraftInsentient.this.lookAt(entity);
        }

        /**
         * Resets the current looking
         */
        @Override
        public void reset() {
            handle.getControllerLook().a();
        }

        /**
         * Checks if the looking is reset
         *
         * @return if looking is default
         */
        @Override
        public boolean isReset() {
            return handle.getControllerLook().b();
        }

        /**
         * Gets the X-Location the entity is looking to
         *
         * @return x-location
         */
        @Override
        public double getLocationX() {
            return handle.getControllerMove().c();
        }

        /**
         * Gets the Y-Location the entity is looking to
         *
         * @return y-location
         */
        @Override
        public double getLocationY() {
            return handle.getControllerMove().d();
        }

        /**
         * Gets the Z-Location the entity is looking to
         *
         * @return z-location
         */
        @Override
        public double getLocationZ() {
            return handle.getControllerMove().f();
        }
    }

}

File: Pathfinder_1_13/src/main/java/com/github/ysl3000/bukkit/pathfinding/craftbukkit/v1_13_R1/entity/CraftInsentient.kt
package com.github.ysl3000.bukkit.pathfinding.craftbukkit.v1_13_R1.entity

import com.github.ysl3000.bukkit.pathfinding.craftbukkit.v1_13_R1.pathfinding.CraftNavigation
import com.github.ysl3000.bukkit.pathfinding.craftbukkit.v1_13_R1.pathfinding.CraftPathfinderGoalWrapper
import com.github.ysl3000.bukkit.pathfinding.entity.Insentient
import com.github.ysl3000.bukkit.pathfinding.pathfinding.ControllerLook
import com.github.ysl3000.bukkit.pathfinding.pathfinding.Navigation
import net.minecraft.server.v1_13_R1.EntityInsentient
import net.minecraft.server.v1_13_R1.PathfinderGoal
import net.minecraft.server.v1_13_R1.PathfinderGoalSelector
import org.bukkit.Location
import org.bukkit.craftbukkit.v1_13_R1.entity.*
import org.bukkit.entity.*
import org.bukkit.util.Vector
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.*

/**
 * Created by ysl3000
 */
class CraftInsentient private constructor(private val handle: EntityInsentient) : Insentient {
    override fun getLookingAt(): Location =
            Location(
                    handle.bukkitEntity.world,
                    handle.controllerLook.e(),
                    handle.controllerLook.f(),
                    handle.controllerLook.g()
            )

    override fun getControllerLook(): ControllerLook = controllerLook
    override fun getControllerMove(): com.github.ysl3000.bukkit.pathfinding.pathfinding.ControllerMove = controllerMove
    override fun getNavigation(): Navigation = navigation

    override fun getTargetSelector(): com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoalSelector = targetSelector
    override fun getGoalSelector(): com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoalSelector = goalSelector

    override fun getControllerJump(): com.github.ysl3000.bukkit.pathfinding.pathfinding.ControllerJump = controllerJump

    override fun getHeadHeight(): Float = handle.headHeight

    override fun getDefaultYaw(): Int = handle.L()

    override fun getDefaultPitch(): Int = handle.K()

    override fun getBukkitEntity(): Entity = handle.bukkitEntity


    private val nmsGoals = HashMap<com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal, PathfinderGoal>()
    private val nmsTargetGoals = HashMap<com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal, PathfinderGoal>()

    constructor(flying: Flying) : this((flying as CraftFlying).handle)

    private val navigation: Navigation
    private val controllerLook: ControllerLook
    private val controllerMove: ControllerMove
    private val controllerJump: ControllerJump
    private val targetSelector: GoalSelector
    private val goalSelector: GoalSelector

    init {
        navigation = CraftNavigation(handle.navigation)
        controllerLook = ControllerLookImpl()
        controllerMove = ControllerMove()
        controllerJump = ControllerJump()
        targetSelector = GoalSelector()
        goalSelector = GoalSelector()
    }


    constructor(enderDragon: EnderDragon) : this((enderDragon as CraftEnderDragon).handle)

    constructor(creature: Creature) : this((creature as CraftCreature).handle)

    constructor(mob: Mob) : this((mob as CraftMob).handle)

    constructor(ambient: Ambient) : this((ambient as CraftAmbient).handle)

    constructor(slime: Slime) : this((slime as CraftSlime).handle)

    override fun addPathfinderGoal(priority: Int,
                                   pathfinderGoal: com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal) {
        val goalWrapper = CraftPathfinderGoalWrapper(pathfinderGoal)
        this.nmsGoals[pathfinderGoal] = goalWrapper
        handle.goalSelector.a(priority, goalWrapper)
    }

    override fun removePathfinderGoal(
            pathfinderGoal: com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal) {
        if (nmsGoals.containsKey(pathfinderGoal)) {
            val nmsGoal = nmsGoals.remove(pathfinderGoal)
            handle.goalSelector.a(nmsGoal)
        }
    }

    override fun hasPathfinderGoal(
            pathfinderGoal: com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal): Boolean {
        return nmsGoals.containsKey(pathfinderGoal)
    }

    override fun clearPathfinderGoals() {
        handle.goalSelector = PathfinderGoalSelector(handle.getWorld().methodProfiler)
        nmsGoals.clear()
    }


    override fun addTargetPathfinderGoal(priority: Int,
                                         pathfinderGoal: com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal) {
        val goalWrapper = CraftPathfinderGoalWrapper(pathfinderGoal)
        this.nmsTargetGoals[pathfinderGoal] = goalWrapper
        handle.targetSelector.a(priority, goalWrapper)
    }

    override fun removeTargetPathfinderGoal(
            pathfinderGoal: com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal) {
        if (nmsTargetGoals.containsKey(pathfinderGoal)) {
            val nmsGoal = nmsTargetGoals.remove(pathfinderGoal)
            handle.goalSelector.a(nmsGoal)
        }
    }

    override fun hasTargetPathfinderGoal(
            pathfinderGoal: com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal): Boolean {
        return nmsTargetGoals.containsKey(pathfinderGoal)
    }

    override fun clearTargetPathfinderGoals() {
        handle.targetSelector = PathfinderGoalSelector(handle.getWorld().methodProfiler)
        nmsTargetGoals.clear()
    }

    override fun jump() {
        handle.controllerJump.a()
    }

    override fun lookAt(location: Location) {
        handle.controllerLook
                .a(location.x, location.y, location.z, location.yaw,
                        location.pitch)
    }

    override fun lookAt(entity: Entity) {
        lookAt(entity.location)
    }

    override fun setMovementDirection(direction: Vector, speed: Double) {
        handle.controllerMove.a(direction.x, direction.blockY.toDouble(), direction.z, speed)
    }

    override fun setStrafeDirection(forward: Float, sideward: Float) {
        handle.controllerMove.a(forward, sideward)
    }

    override fun resetGoalsToDefault() {
        if (reset == null) {
            return
        }
        try {
            reset!!.invoke(handle)
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        }

    }

    override fun hasPositionChanged(): Boolean {
        return handle.positionChanged
    }

    override fun onEntityKill(livingEntity: LivingEntity) {
        handle.b((livingEntity as CraftLivingEntity).handle)
    }

    override fun moveRelative(motionX: Double, motionY: Double, motionZ: Double) {
        moveRelative(motionX, motionY, motionZ, 1.0)
    }

    override fun moveRelative(motionX: Double, motionY: Double, motionZ: Double, speed: Double) {
        this.handle.motX += motionX * speed
        this.handle.motY += motionY * speed
        this.handle.motZ += motionZ * speed
        this.handle.recalcPosition()
    }

    override fun setRotation(yaw: Float, pitch: Float) {
        this.handle.yaw = yaw
        this.handle.pitch = pitch
    }

    override fun updateRenderAngles() {
        handle.controllerMove.a()
    }


    private inner class GoalSelector : com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoalSelector {

        /**
         * Add a pathfinder goal to the entity
         *
         * @param priority the priority 0 highest
         * @param goal The goal to add
         */
        override fun addPathfinderGoal(priority: Int,
                                       goal: com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal) {
            this@CraftInsentient.addPathfinderGoal(priority, goal)
        }

        /**
         * Remove a pathfinder goal from the entity
         *
         * @param goal The goal to remove
         */
        override fun removePathfinderGoal(
                goal: com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal) {
            this@CraftInsentient.removePathfinderGoal(goal)
        }

        /**
         * Will clear all goals
         */
        override fun clearGoals() {
            this@CraftInsentient.clearPathfinderGoals()
        }
    }

    private inner class TargetGoalSelector : com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoalSelector {

        /**
         * Add a pathfinder goal to the entity
         *
         * @param priority the priority 0 highest
         * @param goal The goal to add
         */
        override fun addPathfinderGoal(priority: Int,
                                       goal: com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal) {
            this@CraftInsentient.addTargetPathfinderGoal(priority, goal)
        }

        /**
         * Remove a pathfinder goal from the entity
         *
         * @param goal The goal to remove
         */
        override fun removePathfinderGoal(
                goal: com.github.ysl3000.bukkit.pathfinding.pathfinding.PathfinderGoal) {
            this@CraftInsentient.removeTargetPathfinderGoal(goal)
        }

        /**
         * Will clear all goals
         */
        override fun clearGoals() {
            this@CraftInsentient.clearTargetPathfinderGoals()
        }
    }

    private inner class ControllerJump : com.github.ysl3000.bukkit.pathfinding.pathfinding.ControllerJump {

        /**
         * Lets the entity jump
         */
        override fun jump() {
            this@CraftInsentient.jump()
        }
    }

    private inner class ControllerMove : com.github.ysl3000.bukkit.pathfinding.pathfinding.ControllerMove {
        override fun getX(): Double = handle.controllerMove.c()

        override fun getY(): Double = handle.controllerMove.d()

        override fun getZ(): Double = handle.controllerMove.e()

        override fun getSpeed(): Double = handle.controllerMove.f()

        override fun isOperationMove(): Boolean = handle.controllerMove.b()

        override fun move(motionX: Double, motionY: Double, motionZ: Double, speed: Double) =
                this@CraftInsentient.setMovementDirection(Vector(motionX, motionY, motionZ), speed)

        override fun move(forward: Float, sideward: Float) =
                this@CraftInsentient.setStrafeDirection(forward, sideward)

        override fun update() =
                handle.controllerMove.a()

    }

    private inner class ControllerLookImpl : com.github.ysl3000.bukkit.pathfinding.pathfinding.ControllerLook {

        /**
         * Checks if the looking is reset
         *
         * @return if looking is default
         */
        override fun isReset(): Boolean = handle.controllerLook.b()

        /**
         * Gets the X-Location the entity is looking to
         *
         * @return x-location
         */
        override fun getLocationX(): Double = handle.controllerMove.c()

        /**
         * Gets the Y-Location the entity is looking to
         *
         * @return y-location
         */
        override fun getLocationY(): Double = handle.controllerMove.d()

        /**
         * Gets the Z-Location the entity is looking to
         *
         * @return z-location
         */
        override fun getLocationZ(): Double = handle.controllerMove.f()

        /**
         * The entity will look to the given coordinates
         *
         * @param x x-location
         * @param y y-location
         * @param z z-location
         * @param yaw head rotation yaw
         * @param pitch head rotation pitch
         */
        override fun lookAt(x: Double, y: Double, z: Double, yaw: Float, pitch: Float) = this@CraftInsentient
                .lookAt(Location(handle.bukkitEntity.world, x, y, z, yaw, pitch))

        /**
         * The entity will look to the given location
         *
         * @param location the entity should look to
         */
        override fun lookAt(location: Location) {
            this@CraftInsentient.lookAt(location)
        }

        /**
         * The entity will look to the given entity
         *
         * @param entity the entity which is targeted with eyes
         * @param yaw head rotation yaw
         * @param pitch head rotation pitch
         */
        override fun lookAt(entity: Entity, yaw: Float, pitch: Float) {
            this@CraftInsentient.lookAt(entity)
        }

        /**
         * Resets the current looking
         */
        override fun reset() {
            handle.controllerLook.a()
        }
    }

    companion object {


        private var reset: Method? = null

        init {

            try {
                reset = EntityInsentient::class.java.getDeclaredMethod("n")
                reset!!.isAccessible = true
            } catch (e: NoSuchMethodException) {
                e.printStackTrace()
            }

        }
    }

}
